account/signer.go

```diff
@@ -0,0 +1,115 @@
+package account
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/NethermindEth/starknet.go/curve"
+	"github.com/NethermindEth/starknet.go/utils"
+)
+
+type Signer struct {
+	keystore  *MemKeystore
+	publicKey string
+}
+
+// NewSigner creates a new Signer instance with the provided private key.
+//
+// Parameters:
+// - privateKey: a string representing the private key (in any format).
+//
+// Returns:
+// - *Signer: a pointer to the created Signer instance.
+// - error: an error if the private key is invalid or if the public key cannot be derived.
+func NewSigner(privateKey string) (*Signer, error) {
+	pkBigInt, ok := new(big.Int).SetString(privateKey, 0)
+	if !ok {
+		return nil, fmt.Errorf("invalid private key value")
+	}
+
+	pubKey, err := getPublicKey(pkBigInt)
+	if err != nil {
+		return nil, err
+	}
+
+	keyStore := SetNewMemKeystore(pubKey, pkBigInt)
+
+	return &Signer{
+		keystore:  keyStore,
+		publicKey: pubKey,
+	}, nil
+}
+
+// PublicKey returns the public key associated with the Signer.
+//
+// Returns:
+// - string: the public key.
+func (s *Signer) PublicKey() string {
+	return s.publicKey
+}
+
+// MemKeyStore returns the keystore used by the Signer.
+//
+// Returns:
+// - *MemKeystore: a pointer to the keystore.
+func (s *Signer) MemKeyStore() *MemKeystore {
+	return s.keystore
+}
+
+// Put stores a new private key in the keystore.
+//
+// Parameters:
+// - priv: a string representing the private key in hexadecimal format.
+// - existingKeystore: a pointer to an existing MemKeystore to use. If nil, the Signer's keystore is used.
+//
+// Returns:
+// - error: an error if the private key is invalid.
+func (s *Signer) Put(priv string, existingKeystore *MemKeystore) error {
+	privateKey, ok := new(big.Int).SetString(priv, 16)
+	if !ok {
+		return fmt.Errorf("invalid private key value")
+	}
+
+	var keystoreToUse *MemKeystore
+	if existingKeystore != nil {
+		keystoreToUse = existingKeystore
+	} else {
+		keystoreToUse = s.keystore
+	}
+
+	keystoreToUse.Put(s.publicKey, privateKey)
```

Comment:
The previous publicKey is incorrect here since the user is adding a new privateKey. We must first derive the publicKey from it, assign it to "s.publicKey", and then assign the pubK & privK with the "keystoreToUse.Put"

Reply:
done